e2e/logs_catcher/matcher.go

```diff
@@ -59,37 +61,88 @@ func StartCondition(pctx context.Context, logger *zap.Logger, condition []string
+		}
+	default:
+		return fmt.Errorf("unknown attestation type: %s", attestationType)
+	}
 
-func matchMessages(ctx context.Context, logger *zap.Logger, cli DockerCLI, first []string, second []string, plus int) error {
-	res, err := docker.DockerLogs(ctx, cli, firstTarget, "")
+	// Extract and count logs from the beaconProxyContainer based on the specified criteria.
+	beaconCounts, err := dockerLogsByPubKey(ctx, logger, dockerCLI, beaconProxyContainer, beaconCriteria)
 	if err != nil {
 		return err
 	}
 
-	grepped := res.Grep(first)
-
-	logger.Info("matched", zap.Int("count", len(grepped)), zap.String("target", firstTarget), zap.Strings("match_string", first))
-
-	for _, target := range secondTargets {
-		logger.Debug("Reading one of second targets logs", zap.String("target", target))
-
-		tres, err := docker.DockerLogs(ctx, cli, target, "")
+	// Verify corresponding logs in each SSV node container match the criteria.
+	for _, nodeContainer := range ssvNodesContainers {
+		nodeCounts, err := dockerLogsByPubKey(ctx, logger, dockerCLI, nodeContainer, nodeCriteria)
 		if err != nil {
 			return err
 		}
 
-		tgrepped := tres.Grep(second)
+		// Compare the counts for each public key between beacon proxy and node container.
+		for pubkey, beaconCount := range beaconCounts {
+			nodeCount, exists := nodeCounts[pubkey]
+			if !exists || discrepancyCheck(beaconCount, nodeCount) {
+				logger.Info("Discrepancy found", zap.String("PublicKey", pubkey), zap.Int("BeaconCount", beaconCount), zap.Int("NodeCount", nodeCount))
+				return fmt.Errorf("discrepancy for pubkey %s in %s: expected %d, got %d", pubkey, nodeContainer, beaconCount, nodeCount)
+			}
+		}
+	}
 
-		if len(tgrepped) != len(grepped)+plus {
-			return fmt.Errorf("found non matching messages on %v, want %v got %v", target, len(grepped), len(tgrepped))
+	return nil
+}
+
+// Combines the docker log retrieval, grepping, and counting of public keys into one function.
+func dockerLogsByPubKey(ctx context.Context, logger *zap.Logger, cli DockerCLI, containerName string, matchStrings []string) (map[string]int, error) {
+	res, err := docker.DockerLogs(ctx, cli, containerName, "")
+	if err != nil {
+		return nil, err
+	}
+
+	grepped := res.Grep(matchStrings)
+	logger.Info("matched", zap.Int("count", len(grepped)), zap.String("target", containerName), zap.Strings("match_string", matchStrings))
+	publicKeyCounts := make(map[string]int)
+
+	for _, logStr := range grepped {
+		trimmedLogStr := strings.TrimLeftFunc(logStr, func(r rune) bool {
+			return !strings.ContainsRune("{[", r)
+		})
+
+		var logEntry map[string]interface{}
+		if err := json.Unmarshal([]byte(trimmedLogStr), &logEntry); err != nil {
+			continue // Consider logging this error.
 		}
 
-		logger.Debug("found matching messages for target", zap.Strings("first", first), zap.Strings("second", second), zap.Int("count", len(tgrepped)), zap.String("target", target))
+		if pubkey, ok := logEntry["pubkey"].(string); ok {
```

Comment:
either extract the json parsing code, or use the code already exists in `logs` package for this